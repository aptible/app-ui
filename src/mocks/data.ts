import { defaultBillingDetailResponse } from "@app/billing";
import {
  defaultActivePlanResponse,
  defaultAppResponse,
  defaultBackupRpResponse,
  defaultConfigurationResponse,
  defaultDatabaseImageResponse,
  defaultDatabaseResponse,
  defaultDeployDiskResponse,
  defaultEndpointResponse,
  defaultEnvResponse,
  defaultOperationResponse,
  defaultPermissionResponse,
  defaultPlanResponse,
  defaultServiceResponse,
  defaultStackResponse,
} from "@app/deploy";
import { defaultCodeScanResponse } from "@app/deploy/code-scan-result";
import { createEnv } from "@app/env";
import { defaultHalHref } from "@app/hal";
import { defaultInvitationResponse } from "@app/invitations";
import { defaultOrgResponse } from "@app/organizations";
import { defaultRoleResponse } from "@app/roles";
import { defaultSshKeyResponse } from "@app/ssh-keys";
import { defaultTokenResponse } from "@app/token";
import { defaultUserResponse } from "@app/users";

const idFactory = () => {
  let id = 1;
  return () => {
    id += 1;
    return id;
  };
};
export const createId = idFactory();
export const createText = (mixin: string, id: string | number = "1") => {
  return `test-${mixin}-${id}`;
};

export const testEnv = createEnv({
  origin: "app",
  authUrl: "https://auth.aptible-test.com",
  apiUrl: "https://api.aptible-test.com",
  billingUrl: "https://billing.aptible-test.com",
  legacyDashboardUrl: "https://dashboard.aptible-test.com",
  metricTunnelUrl: "https://metrictunnel.aptible-test.com",
});

export const testUserId = createId();
export const mockJwtHeaders = btoa(
  JSON.stringify({ alg: "HS256", typ: "JWT" }),
);
const mockJwtPayload = (scope: string) =>
  btoa(JSON.stringify({ id: `${testUserId}`, scope }));
export const mockJwt = (mixin: string, scope: string) =>
  `${mockJwtHeaders}.${mockJwtPayload(scope)}.not_real_${mixin}`;

export const testToken = defaultTokenResponse({
  access_token: `${mockJwt(createId().toString(), "manage")}`,
  id: `${createId()}`,
  _links: {
    self: defaultHalHref(),
    user: defaultHalHref(`${testEnv.authUrl}/users/${testUserId}`),
    actor: defaultHalHref(`${testEnv.authUrl}/users/${testUserId}`),
  },
});

export const testElevatedToken = defaultTokenResponse({
  access_token: `${mockJwt(createId().toString(), "elevated")}`,
  scope: "elevated",
  id: `${createId()}`,
  _links: {
    self: defaultHalHref(),
    user: defaultHalHref(`${testEnv.authUrl}/users/${testUserId}`),
    actor: defaultHalHref(`${testEnv.authUrl}/users/${testUserId}`),
  },
});

export const testEmail = "test@aptible.com";
export const testUser = defaultUserResponse({
  id: testUserId,
  email: testEmail,
  verified: false,
});
export const testUserVerified = defaultUserResponse({
  id: testUserId,
  email: "test.verified@aptible.com",
  verified: true,
});
export const testUserVerifiedSecond = defaultUserResponse({
  id: testUserId,
  email: "test.verified.2@aptible.com",
  verified: true,
});

export const testSshKey = defaultSshKeyResponse({ id: `${createId()}` });

export const testBillingDetail = defaultBillingDetailResponse({
  id: `${createId()}`,
  _links: {
    payment_method: defaultHalHref(
      `${testEnv.billingUrl}/external_payment_sources/404`,
    ),
  },
});

export const testOrg = defaultOrgResponse({
  name: createText("org"),
  id: `${createId()}`,
  _links: {
    billing_detail: defaultHalHref(
      `${testEnv.billingUrl}/billing_details/${testBillingDetail.id}`,
    ),
    self: defaultHalHref(),
    invitations: defaultHalHref(),
    roles: defaultHalHref(),
    security_officer: defaultHalHref(),
    users: defaultHalHref(),
  },
});
export const testOrgSpecial = defaultOrgResponse({
  id: `${createId()}`,
  name: "Wow Org",
  updated_at: new Date("2023-01-01").toISOString(),
});
export const testOrgReauth = defaultOrgResponse({
  id: `${createId()}`,
  name: "Reauth Org",
  updated_at: new Date("2023-01-01").toISOString(),
  reauth_required: true,
});
export const testUserOrgSelected = defaultUserResponse({
  id: testUserId,
  email: "special@aptible.com",
  verified: true,
  _links: {
    current_otp_configuration: defaultHalHref(),
    email_verification_challenges: defaultHalHref(),
    roles: defaultHalHref(),
    self: defaultHalHref(),
    ssh_keys: defaultHalHref(),
    u2f_devices: defaultHalHref(),
    selected_organization: defaultHalHref(
      `${testEnv.authUrl}/organizations/${testOrgSpecial.id}`,
    ),
  },
});

export const testStack = defaultStackResponse({
  id: createId(),
  name: createText("stack"),
  region: "us-east-1",
});

export const testRole = defaultRoleResponse({
  id: `${createId()}`,
  name: "Deploy User",
  type: "platform_user",
  _links: {
    organization: defaultHalHref(testOrg.id),
  },
});

export const testRoleOwner = defaultRoleResponse({
  id: `${createId()}`,
  name: "Deploy Owner",
  type: "platform_owner",
  _links: {
    organization: defaultHalHref(testOrg.id),
  },
});

const testAccountId = createId();
export const testAccount = defaultEnvResponse({
  id: testAccountId,
  handle: createText("account"),
  organization_id: testOrg.id,
  _links: {
    stack: defaultHalHref(`${testEnv.apiUrl}/stacks/${testStack.id}`),
  },
  _embedded: {
    permissions: [
      defaultPermissionResponse({
        id: `${createId()}`,
        scope: "deploy",
        _links: {
          account: defaultHalHref(
            `${testEnv.apiUrl}/accounts/${testAccountId}`,
          ),
          role: defaultHalHref(`${testEnv.apiUrl}/roles/${testRole.id}`),
        },
      }),
    ],
  },
});

export const testAccountAdmin = defaultEnvResponse({
  id: testAccountId,
  handle: createText("account"),
  organization_id: testOrg.id,
  _links: {
    stack: defaultHalHref(`${testEnv.apiUrl}/stacks/${testStack.id}`),
  },
  _embedded: {
    permissions: [
      defaultPermissionResponse({
        id: `${createId()}`,
        scope: "admin",
        _links: {
          account: defaultHalHref(
            `${testEnv.apiUrl}/accounts/${testAccountId}`,
          ),
          role: defaultHalHref(`${testEnv.apiUrl}/roles/${testRole.id}`),
        },
      }),
    ],
  },
});

const testDestroyAccountId = createId();
export const testDestroyAccount = defaultEnvResponse({
  id: testDestroyAccountId,
  handle: createText("account-destroy"),
  organization_id: testOrg.id,
  _links: {
    stack: defaultHalHref(`${testEnv.apiUrl}/stacks/${testStack.id}`),
  },
  _embedded: {
    permissions: [
      defaultPermissionResponse({
        id: `${createId()}`,
        scope: "destroy",
        _links: {
          account: defaultHalHref(
            `${testEnv.apiUrl}/accounts/${testDestroyAccountId}`,
          ),
          role: defaultHalHref(`${testEnv.apiUrl}/roles/${testRole.id}`),
        },
      }),
    ],
  },
});

const testAppId = createId();

export const testServiceRails = defaultServiceResponse({
  id: createId(),
  handle: createText("web"),
  command: "rails s",
  container_count: 1,
  container_memory_limit_mb: 512,
  instance_class: "m5",
  _links: {
    current_release: defaultHalHref(),
    app: defaultHalHref(`${testEnv.apiUrl}/apps/${testAppId}`),
    account: defaultHalHref(`${testEnv.apiUrl}/accounts/${testAccountId}`),
  },
});
export const testServiceSidekiq = defaultServiceResponse({
  id: createId(),
  handle: createText("background"),
  command: "rake sidekiq",
  container_count: 1,
  container_memory_limit_mb: 512,
  instance_class: "m5",
  _links: {
    current_release: defaultHalHref(),
    app: defaultHalHref(`${testEnv.apiUrl}/apps/${testAppId}`),
    account: defaultHalHref(`${testEnv.apiUrl}/accounts/${testAccountId}`),
  },
});

export const testConfiguration = defaultConfigurationResponse({
  id: createId(),
  env: { DATABASE_URL: "{{test-app-1-postgres}}" },
  _links: {
    resource: defaultHalHref(`${testEnv.apiUrl}/apps/${testAppId}`),
  },
});

export const testApp = defaultAppResponse({
  id: testAppId,
  handle: createText("app"),
  _links: {
    account: defaultHalHref(`${testEnv.apiUrl}/accounts/${testAccount.id}`),
    current_configuration: defaultHalHref(),
    current_image: defaultHalHref(),
  },
  _embedded: {
    current_image: null,
    last_operation: null,
    last_deploy_operation: null,
    services: [testServiceRails, testServiceSidekiq],
  },
});

export const testScanOperation = defaultOperationResponse({
  id: createId(),
  type: "scan_code",
  status: "succeeded",
  _links: {
    code_scan_result: defaultHalHref(
      `${testEnv.apiUrl}/code_scan_results/${createId()}`,
    ),
    resource: defaultHalHref(`${testEnv.apiUrl}/apps/${testApp.id}`),
    ephemeral_sessions: defaultHalHref(),
    self: defaultHalHref(),
    account: testApp._links.account,
    ssh_portal_connections: defaultHalHref(),
    user: defaultHalHref(),
    logs: defaultHalHref(),
  },
});

export const testCodeScanResult = defaultCodeScanResponse({
  id: createId(),
  dockerfile_present: true,
  _links: {
    app: defaultHalHref(`${testEnv.apiUrl}/apps/${testApp.id}`),
    operation: defaultHalHref(),
  },
});

export const testPostgresDatabaseImage = defaultDatabaseImageResponse({
  id: createId(),
  type: "postgres",
  version: "14",
  description: "postgres v14",
});

export const testRedisDatabaseImage = defaultDatabaseImageResponse({
  id: createId(),
  type: "redis",
  version: "5",
  description: "redis v5",
});

export const testDatabaseId = createId();
export const testDatabaseServiceId = createId();
export const testDatabaseOp = defaultOperationResponse({
  id: createId(),
  type: "provision",
  status: "succeeded",
  _links: {
    resource: defaultHalHref(`${testEnv.apiUrl}/databases/${testDatabaseId}`),
    account: defaultHalHref(`${testEnv.apiUrl}/accounts/${testAccount.id}`),
    code_scan_result: defaultHalHref(),
    self: defaultHalHref(),
    ssh_portal_connections: defaultHalHref(),
    ephemeral_sessions: defaultHalHref(),
    logs: defaultHalHref(),
    user: defaultHalHref(),
  },
});

export const testDisk = defaultDeployDiskResponse({
  id: `${createId()}`,
  size: 10,
});

export const testDatabasePostgres = defaultDatabaseResponse({
  id: testDatabaseId,
  handle: `${testApp.handle}-postgres`,
  type: "postgres",
  connection_url: "postgres://some:val@wow.com:5432",
  _embedded: {
    disk: testDisk,
    last_operation: testDatabaseOp,
  },
  _links: {
    account: defaultHalHref(`${testEnv.apiUrl}/accounts/${testAccount.id}`),
    initialize_from: defaultHalHref(),
    database_image: defaultHalHref(
      `${testEnv.apiUrl}/database_images/${testPostgresDatabaseImage.id}`,
    ),
    service: defaultHalHref(
      `${testEnv.apiUrl}/services/${testDatabaseServiceId}`,
    ),
    disk: defaultHalHref(`${testEnv.apiUrl}/disks/${testDisk.id}`),
  },
});

export const testDatabaseInfluxdb = defaultDatabaseResponse({
  id: testDatabaseId,
  handle: "influxdb-example",
  type: "influxdb",
  _links: {
    account: defaultHalHref(`${testEnv.apiUrl}/accounts/${testAccount.id}`),
    initialize_from: defaultHalHref(),
    database_image: defaultHalHref(),
    service: defaultHalHref(),
    disk: defaultHalHref(`${testEnv.apiUrl}/disks/${testDisk.id}`),
  },
});

export const testDatabaseElasticsearch = defaultDatabaseResponse({
  id: testDatabaseId,
  handle: "elasticsearch-example",
  type: "elasticsearch",
  _links: {
    account: defaultHalHref(`${testEnv.apiUrl}/accounts/${testAccount.id}`),
    initialize_from: defaultHalHref(),
    database_image: defaultHalHref(),
    service: defaultHalHref(),
    disk: defaultHalHref(`${testEnv.apiUrl}/disks/${testDisk.id}`),
  },
});

export const testServicePostgres = defaultServiceResponse({
  id: testDatabaseServiceId,
  handle: createText("postgres"),
  command: undefined,
  container_count: 1,
  container_memory_limit_mb: 512,
  process_type: "postgresql",
  _links: {
    current_release: defaultHalHref(),
    database: defaultHalHref(`${testEnv.apiUrl}/databases/${testDatabaseId}`),
    account: defaultHalHref(`${testEnv.apiUrl}/accounts/${testAccount.id}`),
  },
});

export const testEndpoint = defaultEndpointResponse({
  id: createId(),
  _links: {
    service: defaultHalHref(
      `${testEnv.apiUrl}/services/${testServiceRails.id}`,
    ),
    certificate: defaultHalHref(),
  },
});

export const testPlan = defaultPlanResponse({
  id: createId(),
});
export const testEnterprisePlan = defaultPlanResponse({
  id: createId(),
  name: "enterprise",
});

export const testActivePlan = defaultActivePlanResponse({
  id: createId(),
  available_plans: ["growth", "scale"],
  organization_id: testOrg.id,
  _links: {
    organization: defaultHalHref(
      `${testEnv.authUrl}/organizations/${testOrg.id}`,
    ),
    plan: defaultHalHref(`${testEnv.apiUrl}/plans/${testPlan.id}`),
  },
});

export const testEnvExpress = defaultEnvResponse({
  id: createId(),
  organization_id: testOrg.id,
  handle: createText("express"),
  onboarding_status: "initiated",
  _links: {
    stack: defaultHalHref(`${testEnv.apiUrl}/stacks/${testStack.id}`),
  },
});
export const testAppDeployedId = createId();
const testAppDeployOperation = defaultOperationResponse({
  id: createId(),
  type: "deploy",
  status: "succeeded",
  updated_at: new Date("2023-04-08T14:00:00.0000").toISOString(),
  _links: {
    resource: defaultHalHref(`${testEnv.apiUrl}/apps/${testAppDeployedId}`),
    account: defaultHalHref(`${testEnv.apiUrl}/accounts/${testEnvExpress.id}`),
    code_scan_result: defaultHalHref(),
    ephemeral_sessions: defaultHalHref(),
    logs: defaultHalHref(),
    ssh_portal_connections: defaultHalHref(),
    self: defaultHalHref(),
    user: defaultHalHref(),
  },
});
export const testAppDeployed = defaultAppResponse({
  id: testAppDeployedId,
  handle: `${testEnvExpress.handle}-app`,
  _links: {
    account: defaultHalHref(`${testEnv.apiUrl}/accounts/${testEnvExpress.id}`),
    current_configuration: defaultHalHref(),
    current_image: defaultHalHref(),
  },
  _embedded: {
    services: [],
    current_image: null,
    last_operation: null,
    last_deploy_operation: testAppDeployOperation,
  },
});

export const testOperations = [
  testScanOperation,
  testDatabaseOp,
  testAppDeployOperation,
];

export const testBackupRp = defaultBackupRpResponse({
  id: createId(),
  daily: 1,
  monthly: 5,
  keep_final: true,
  make_copy: false,
  _links: {
    account: defaultHalHref(`${testEnv.apiUrl}/accounts/${testAccountId}`),
  },
});

export const testVerifiedInvitation = defaultInvitationResponse({
  id: `${createId()}`,
  email: testUserVerified.email,
  organization_name: testOrg.name,
  inviter_name: "test.owner@aptible.com",
  role_name: testRole.name,
  _links: {
    organization: defaultHalHref(
      `${testEnv.authUrl}/organizations/${testOrg.id}`,
    ),
  },
});
