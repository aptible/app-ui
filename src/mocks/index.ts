export * from "./server";
export * from "./data";
export * from "./handlers";
