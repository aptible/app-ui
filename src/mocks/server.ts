import { handlers } from "./handlers";
// src/mocks/server.js
import { setupServer } from "msw/node";

// This configures a request mocking server with the given request handlers.
export const server = setupServer(...handlers);
