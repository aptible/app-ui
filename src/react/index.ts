import { selectLoaderComposite } from "@app/loaders";
import { type WebState, schema } from "@app/schema";
import { useEffect, useRef } from "react";
import type { LoaderState, ThunkAction } from "starfx";
import {
  type TypedUseSelectorHook,
  useSelector as useBaseSelector,
  useDispatch,
} from "starfx/react";
export const useSelector: TypedUseSelectorHook<WebState> = useBaseSelector;

export {
  useDispatch,
  PersistGate,
  Provider,
} from "starfx/react";

type ActionFn<P = any> = (p: P) => { toString: () => string };
type ActionFnSimple = () => { toString: () => string };

export interface UseApiProps<P = any> extends LoaderState {
  trigger: (p: P) => void;
  action: ActionFn<P>;
}
export interface UseApiSimpleProps extends LoaderState {
  trigger: () => void;
  action: ActionFn;
}
export interface UseApiAction<A extends ThunkAction = ThunkAction>
  extends LoaderState {
  trigger: () => void;
  action: A;
}
export type UseApiResult<P, A extends ThunkAction = ThunkAction> =
  | UseApiProps<P>
  | UseApiSimpleProps
  | UseApiAction<A>;

interface UseCacheResult<D = any, A extends ThunkAction = ThunkAction>
  extends UseApiAction<A> {
  data: D | null;
}

function getActionId(action: ThunkAction | ActionFn) {
  return typeof action === "function" ? `${action}` : action.payload.key;
}

export function useLoader(action: ThunkAction | ActionFn) {
  const id = getActionId(action);
  return useSelector((s: WebState) => schema.loaders.selectById(s, { id }));
}

export function useCompositeLoader(actions: (ThunkAction | ActionFn)[]) {
  const ids = actions.map((a) => getActionId(a));
  return useSelector((s) => selectLoaderComposite(s, { ids }));
}

export function useApi<P = any, A extends ThunkAction = ThunkAction<P>>(
  action: A,
): UseApiAction<A>;
export function useApi<P = any, A extends ThunkAction = ThunkAction<P>>(
  action: ActionFn<P>,
): UseApiProps<P>;
export function useApi<A extends ThunkAction = ThunkAction>(
  action: ActionFnSimple,
): UseApiSimpleProps;
export function useApi(action: any): any {
  const dispatch = useDispatch();
  const loader = useLoader(action);
  const trigger = (p: any) => {
    if (typeof action === "function") {
      dispatch(action(p));
    } else {
      dispatch(action);
    }
  };
  return { ...loader, trigger, action };
}

export function useQuery<P = any, A extends ThunkAction = ThunkAction<P>>(
  action: A,
): UseApiAction<A> {
  const api = useApi(action);
  useEffect(() => {
    api.trigger();
  }, [action.payload.key]);
  return api;
}

export function useCache<P = any, ApiSuccess = any>(
  action: ThunkAction<P, ApiSuccess>,
): UseCacheResult<typeof action.payload._result, ThunkAction<P, ApiSuccess>> {
  const id = action.payload.key;
  const data: any = useSelector((s: WebState) =>
    schema.cache.selectById(s, { id }),
  );
  const query = useQuery(action);
  return { ...query, data: data || null };
}

export function useLoaderSuccess(
  cur: Pick<LoaderState, "status">,
  success: () => any,
) {
  const prev = useRef(cur);
  useEffect(() => {
    if (prev.current.status !== "success" && cur.status === "success") {
      success();
    }
    prev.current = cur;
  }, [cur.status]);
}
