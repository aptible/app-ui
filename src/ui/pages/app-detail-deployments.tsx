import { prettyDateTime } from "@app/date";
import { selectAppById, selectOperationById } from "@app/deploy";
import {
  fetchDeploymentsByAppId,
  selectDeploymentsByAppId,
} from "@app/deployment";
import { deploymentDetailRollbackUrl, deploymentDetailUrl } from "@app/routes";
import { AppState, Deployment } from "@app/types";
import { useSelector } from "react-redux";
import { Link, useParams } from "react-router-dom";
import { useQuery } from "starfx/react";
import {
  ButtonLink,
  Code,
  DetailPageSections,
  OpStatus,
  TBody,
  THead,
  Table,
  Td,
  Tr,
} from "../shared";

function DeploymentRow({ deployment }: { deployment: Deployment }) {
  const op = useSelector((s: AppState) =>
    selectOperationById(s, { id: deployment.operationId }),
  );
  let ref = "";
  if (deployment.dockerImage) {
    ref = deployment.dockerImage.replace(/^[\w.\/-]+:/, "");
  } else if (deployment.gitRef) {
    ref = deployment.gitRef;
  }

  return (
    <Tr>
      <Td>
        <Link to={deploymentDetailUrl(deployment.id)}>{deployment.id}</Link>
      </Td>
      <Td>
        <OpStatus status={op.status} />
      </Td>
      <Td>{op.type}</Td>
      <Td>
        <Code>{ref}</Code>
      </Td>
      <Td>{prettyDateTime(deployment.createdAt)}</Td>
      <Td variant="right">
        <ButtonLink size="sm" to={deploymentDetailRollbackUrl(deployment.id)}>
          Rollback
        </ButtonLink>
      </Td>
    </Tr>
  );
}

export function AppDetailDeploymentsPage() {
  const { id = "" } = useParams();
  const app = useSelector((s: AppState) => selectAppById(s, { id }));
  useQuery(fetchDeploymentsByAppId({ id: app.id }));
  const deployments = useSelector((s: AppState) =>
    selectDeploymentsByAppId(s, { appId: app.id }),
  );

  return (
    <DetailPageSections>
      <Table>
        <THead>
          <Td>ID</Td>
          <Td>Status</Td>
          <Td>Type</Td>
          <Td>Ref</Td>
          <Td>Date</Td>
          <Td variant="right">Actions</Td>
        </THead>

        <TBody>
          {deployments.map((deploy) => {
            return <DeploymentRow key={deploy.id} deployment={deploy} />;
          })}
        </TBody>
      </Table>
    </DetailPageSections>
  );
}
