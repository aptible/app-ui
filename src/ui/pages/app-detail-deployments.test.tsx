import { server, stacksWithResources, testApp } from "@app/mocks";
import {
  APP_DETAIL_DEPLOYMENTS_PATH,
  appDetailDeploymentsUrl,
} from "@app/routes";
import { setupIntegrationTest, waitForBootup } from "@app/test";
import { render, screen } from "@testing-library/react";
import { AppDetailDeploymentsPage } from "./app-detail-deployments";

describe("App Detail Deployments page", () => {
  it("should render deployments with proper fallbacks", async () => {
    server.use(...stacksWithResources({ apps: [testApp] }));
    const { store, TestProvider } = setupIntegrationTest({
      initEntries: [appDetailDeploymentsUrl(`${testApp.id}`)],
      path: APP_DETAIL_DEPLOYMENTS_PATH,
    });
    await waitForBootup(store);
    render(
      <TestProvider>
        <AppDetailDeploymentsPage />
      </TestProvider>,
    );

    const links = await screen.findAllByRole("link", { name: "a947a95" });
    expect(links[0].getAttribute("href")).toEqual(
      "https://github.com/aptible/app-ui/commit/a947a95a92e7a7a4db7fe01c28346281c128b859",
    );
    // expect(screen.queryByText(/quay.io\/aptible\/cloud-ui/)).toBeInTheDocument();
  });
});
