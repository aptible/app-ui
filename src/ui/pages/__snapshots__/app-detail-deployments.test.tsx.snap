// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App Detail Deployments page > should render deployments with proper fallbacks 1`] = `
<div>
  <div
    class="flex flex-col"
  >
    <div
      class="overflow-x-scroll w-full shadow ring-1 ring-black ring-opacity-5 rounded-lg"
    >
      <div
        class="min-w-[600px]"
      >
        <table
          class="table-auto w-full divide-y divide-gray-300"
        >
          <thead
            class="bg-gray-50"
          >
            <tr>
              <td
                class="text-sm text-gray-500 pl-4 py-3 break-words text-left"
                colspan="1"
              >
                ID
              </td>
              <td
                class="text-sm text-gray-500 pl-4 py-3 break-words text-left"
                colspan="1"
              >
                Status
              </td>
              <td
                class="text-sm text-gray-500 pl-4 py-3 break-words text-left"
                colspan="1"
              >
                Source
              </td>
              <td
                class="text-sm text-gray-500 pl-4 py-3 break-words text-left"
                colspan="1"
              >
                Tag
              </td>
              <td
                class="text-sm text-gray-500 pl-4 py-3 break-words text-left"
                colspan="1"
              >
                Commit
              </td>
              <td
                class="text-sm text-gray-500 pl-4 py-3 break-words text-left"
                colspan="1"
              >
                Message
              </td>
              <td
                class="text-sm text-gray-500 pl-4 py-3 break-words text-left"
                colspan="1"
              >
                Date
              </td>
              <td
                class="text-sm text-gray-500 pl-4 py-3 break-words text-right pr-4 flex gap-2 justify-end"
                colspan="1"
              >
                Actions
              </td>
            </tr>
          </thead>
          <tbody
            class="divide-y divide-gray-200 bg-white"
          >
            <tr
              class="group hover:bg-gray-50 "
            >
              <td
                class="text-sm text-gray-500 pl-4 py-3 break-words text-left"
                colspan="1"
              >
                <a
                  class="text-black group-hover:text-indigo hover:text-indigo"
                  href="/deployments/1"
                >
                  1
                </a>
              </td>
              <td
                class="text-sm text-gray-500 pl-4 py-3 break-words text-left"
                colspan="1"
              >
                <span
                  class="font-semibold text-red"
                >
                  FAILED
                </span>
              </td>
              <td
                class="text-sm text-gray-500 pl-4 py-3 break-words text-left"
                colspan="1"
              >
                Aptible Git Deployment
              </td>
              <td
                class="text-sm text-gray-500 pl-4 py-3 break-words text-left"
                colspan="1"
              />
              <td
                class="text-sm text-gray-500 pl-4 py-3 break-words text-left"
                colspan="1"
              />
              <td
                class="text-sm text-gray-500 pl-4 py-3 break-words text-left"
                colspan="1"
              >
                <span />
              </td>
              <td
                class="text-sm text-gray-500 pl-4 py-3 break-words text-left"
                colspan="1"
              >
                2023-12-17 08:24:00 UTC
              </td>
              <td
                class="text-sm text-gray-500 pl-4 py-3 break-words text-right pr-4 flex gap-2 justify-end"
                colspan="1"
              >
                <a
                  class="w-15 flex items-center justify-center px-3 py-2 text-sm leading-4 font-medium border border-transparent shadow-sm font-bold text-black bg-gradient-to-r from-yellow-400 to-yellow-500 hover:no-underline hover:from-yellow hover:to-yellow focus:from-orange-400 focus:to-orange-400 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-yellow-500 disabled:from-orange-200 disabled:to-orange-200 disabled:text-black-300 disabled:cursor-not-allowed rounded-md hover:no-underline cursor-pointer"
                  href="/deployments/1"
                >
                  View
                </a>
              </td>
            </tr>
            <tr
              class="group hover:bg-gray-50 "
            >
              <td
                class="text-sm text-gray-500 pl-4 py-3 break-words text-left"
                colspan="1"
              >
                <a
                  class="text-black group-hover:text-indigo hover:text-indigo"
                  href="/deployments/2"
                >
                  2
                </a>
              </td>
              <td
                class="text-sm text-gray-500 pl-4 py-3 break-words text-left"
                colspan="1"
              >
                <span
                  class="font-semibold text-forest"
                >
                  DONE
                </span>
              </td>
              <td
                class="text-sm text-gray-500 pl-4 py-3 break-words text-left"
                colspan="1"
              >
                <a
                  class=" "
                  href="https://github.com/aptible/app-ui"
                  rel="noreferrer"
                  target="_blank"
                >
                  github.com/aptible/app-ui
                </a>
              </td>
              <td
                class="text-sm text-gray-500 pl-4 py-3 break-words text-left"
                colspan="1"
              >
                <code
                  class=" bg-gray-200 font-mono text-black pt-0.5 pb-1 px-1.5 rounded-md text-[0.9rem]"
                >
                  v3
                </code>
              </td>
              <td
                class="text-sm text-gray-500 pl-4 py-3 break-words text-left"
                colspan="1"
              >
                <code
                  class=" bg-gray-200 font-mono text-black pt-0.5 pb-1 px-1.5 rounded-md text-[0.9rem]"
                >
                  <a
                    class=" "
                    href="https://github.com/aptible/app-ui/commit/a947a95a92e7a7a4db7fe01c28346281c128b859"
                    rel="noreferrer"
                    target="_blank"
                  >
                    a947a95
                  </a>
                </code>
              </td>
              <td
                class="text-sm text-gray-500 pl-4 py-3 break-words text-left"
                colspan="1"
              >
                <span>
                  fix(backup): pass page to fetch request (#754)
                </span>
              </td>
              <td
                class="text-sm text-gray-500 pl-4 py-3 break-words text-left"
                colspan="1"
              >
                2023-12-17 08:24:00 UTC
              </td>
              <td
                class="text-sm text-gray-500 pl-4 py-3 break-words text-right pr-4 flex gap-2 justify-end"
                colspan="1"
              >
                <a
                  class="w-15 flex items-center justify-center px-3 py-2 text-sm leading-4 font-medium border border-transparent shadow-sm font-bold text-black bg-gradient-to-r from-yellow-400 to-yellow-500 hover:no-underline hover:from-yellow hover:to-yellow focus:from-orange-400 focus:to-orange-400 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-yellow-500 disabled:from-orange-200 disabled:to-orange-200 disabled:text-black-300 disabled:cursor-not-allowed rounded-md hover:no-underline cursor-pointer"
                  href="/deployments/2"
                >
                  View
                </a>
              </td>
            </tr>
            <tr
              class="group hover:bg-gray-50 "
            >
              <td
                class="text-sm text-gray-500 pl-4 py-3 break-words text-left"
                colspan="1"
              >
                <a
                  class="text-black group-hover:text-indigo hover:text-indigo"
                  href="/deployments/3"
                >
                  3
                </a>
              </td>
              <td
                class="text-sm text-gray-500 pl-4 py-3 break-words text-left"
                colspan="1"
              >
                <span
                  class="font-semibold text-black"
                >
                  PENDING
                </span>
              </td>
              <td
                class="text-sm text-gray-500 pl-4 py-3 break-words text-left"
                colspan="1"
              >
                <a
                  class=" "
                  href="https://github.com/aptible/app-ui"
                  rel="noreferrer"
                  target="_blank"
                >
                  github.com/aptible/app-ui
                </a>
              </td>
              <td
                class="text-sm text-gray-500 pl-4 py-3 break-words text-left"
                colspan="1"
              >
                <code
                  class=" bg-gray-200 font-mono text-black pt-0.5 pb-1 px-1.5 rounded-md text-[0.9rem]"
                >
                  v3
                </code>
              </td>
              <td
                class="text-sm text-gray-500 pl-4 py-3 break-words text-left"
                colspan="1"
              >
                <code
                  class=" bg-gray-200 font-mono text-black pt-0.5 pb-1 px-1.5 rounded-md text-[0.9rem]"
                >
                  <a
                    class=" "
                    href="https://github.com/aptible/app-ui/commit/a947a95a92e7a7a4db7fe01c28346281c128b859"
                    rel="noreferrer"
                    target="_blank"
                  >
                    a947a95
                  </a>
                </code>
              </td>
              <td
                class="text-sm text-gray-500 pl-4 py-3 break-words text-left"
                colspan="1"
              >
                <span>
                  fix(backup): pass page to fetch request (#754)
                </span>
              </td>
              <td
                class="text-sm text-gray-500 pl-4 py-3 break-words text-left"
                colspan="1"
              >
                2023-12-17 08:24:00 UTC
              </td>
              <td
                class="text-sm text-gray-500 pl-4 py-3 break-words text-right pr-4 flex gap-2 justify-end"
                colspan="1"
              >
                <a
                  class="w-15 flex items-center justify-center px-3 py-2 text-sm leading-4 font-medium border border-transparent shadow-sm font-bold text-black bg-gradient-to-r from-yellow-400 to-yellow-500 hover:no-underline hover:from-yellow hover:to-yellow focus:from-orange-400 focus:to-orange-400 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-yellow-500 disabled:from-orange-200 disabled:to-orange-200 disabled:text-black-300 disabled:cursor-not-allowed rounded-md hover:no-underline cursor-pointer"
                  href="/deployments/3"
                >
                  View
                </a>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
